### PR TITLE
Implement IEnumerable<> on various X509 collection types

### LINKED ISSUE #49878: API Proposal: Implement IEnumerable<> on various X509 collection types
## Background and Motivation

X509Certificate2Collection and related types implement the non-generic enumerable but not generic enumerable.  This makes them harder to use with LINQ, and it's very common to see `.OfType<X509Whatever>()` used with these types, even though all of the elements of the collection are guaranteed to implement that type.  This makes consumption both more cumbersome, more confusing, and more expensive.  We can't change the signatures of the public GetEnumerator methods on these types, but we can at least implement the interfaces so that the OfType isn't necessary.

## Proposed API

```diff
public class X509Certificate2Collection : X509CertificateCollection,
+     IEnumerable<X509Certificate2>
{
    ...
}

public sealed class X509Certificate2Enumerator : IEnumerator,
+     IEnumerator<X509Certificate2>
{
    ...
}
    

public sealed class X509ChainElementCollection : ICollection,
+    IEnumerable<X509ChainElement>
{
    ...
}

public sealed class X509ChainElementEnumerator : IEnumerator,
+    IEnumerator<X509ChainElement>
{
    ...
}

public sealed class X509ExtensionCollection : ICollection,
+    IEnumerable<X509Extension>
{
    ...
}

public sealed class X509ExtensionEnumerator : IEnumerator,
+    IEnumerator<X509Extension>
{
    ...
}
```
All new interface method implementations would be explicit.

(We could also consider implementing `ICollection<T>` where the existing type implements `ICollection`.)

Example commit:
https://github.com/stephentoub/runtime/commit/d310e84f5b1918f5886edf640fdfee13cb01393d

### PR DESCRIPTION
**Pull request by hrrrrustic**
close #49878

### CODE REVIEW

**Comment by stephentoub** [comment]
Thanks. Note the linked issue linked to a commit that sketched out the changes (https://github.com/stephentoub/runtime/commit/d310e84f5b1918f5886edf640fdfee13cb01393d), and as part of that, it updated places that were then unnecessarily using OfType... it'd be good to do the same here.  Also, this PR should also be adding tests exercising the new interfaces (or if existing tests end up exercising the new interfaces, that then means they're no longer exercising the old ones, so new tests should be added for those).

File: src/libraries/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/SignedCms.cs [hunk 1]
```diff
@@ -557,7 +557,7 @@ internal void UpdateCertificatesFromAddition(X509Certificate2Collection newCerts
 
             if (existingLength > 0 || newCerts.Count > 1)
             {
-                var certs = new HashSet<X509Certificate2>(Certificates.OfType<X509Certificate2>());
+                var certs = new HashSet<X509Certificate2>(Certificates);
```
Comment (vcsjones) [inline src/libraries/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/SignedCms.cs:560]: The `System.Security.Cryptography.Pkcs` package is still built against .NET Framework (which won't have the new interface implementations).  You won't be able to remove the `OfType` from projects that build against anything other than `NetCoreAppCurrent`

https://github.com/dotnet/runtime/blob/2df392394d3a8d88b451486ea1dd983b70ed9e0d/src/libraries/System.Security.Cryptography.Pkcs/src/System.Security.Cryptography.Pkcs.csproj#L7

It's probably fine to leave the `OfType` and other things in the `Pkcs` package.

**Comment by hrrrrustic** [comment]
I'm a little bit confused with adding tests

All existed tests seems like that
https://github.com/dotnet/runtime/blob/3846a9c4e5529ab1da1d58058f86502da85fead6/src/libraries/System.Security.Cryptography.X509Certificates/tests/CollectionTests.cs#L174-L203

or that
https://github.com/dotnet/runtime/blob/3846a9c4e5529ab1da1d58058f86502da85fead6/src/libraries/System.Security.Cryptography.X509Certificates/tests/CollectionTests.cs#L380-L395

Should I really add exactly same for typed `IEnumerable`/`IEnumerator`?

**Comment by bartonjs** [comment]
> Should I really add exactly same for typed IEnumerable/IEnumerator?

Since it's the same method bodies providing the behavior, nah.

**Review by bartonjs — COMMENTED** [review]
LGTM.  Added Steve to see if he has any remaining thoughts based on the prototype/investigation work.

File: src/libraries/System.Security.Cryptography.X509Certificates/ref/System.Security.Cryptography.X509Certificates.cs [hunk 2]
```diff
@@ -310,6 +311,7 @@ public sealed partial class X509Certificate2Enumerator : System.Collections.IEnu
         public void Reset() { }
         bool System.Collections.IEnumerator.MoveNext() { throw null; }
         void System.Collections.IEnumerator.Reset() { }
+        void System.IDisposable.Dispose() { }
```
Comment (stephentoub) [inline src/libraries/System.Security.Cryptography.X509Certificates/ref/System.Security.Cryptography.X509Certificates.cs:314]: @bartonjs, opinion on this being explicitly implemented vs implicitly?

**Review by stephentoub — APPROVED** [review]
Thanks!

File: src/libraries/System.Security.Cryptography.X509Certificates/ref/System.Security.Cryptography.X509Certificates.cs (hunk 2 above)
Comment (bartonjs) [reply src/libraries/System.Security.Cryptography.X509Certificates/ref/System.Security.Cryptography.X509Certificates.cs:314]: For a sealed type that most people never use directly where it's a no-op merely required by interface inheritance?  Nah.

**Review by bartonjs — APPROVED** [review]

**Pull request merged**